python/shotgun_model/shotgun_entity_model.py

```diff
@@ -0,0 +1,249 @@
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights 
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import sgtk
+from sgtk.platform.qt import QtGui, QtCore
+
+from .shotgun_model import ShotgunModel
+from .util import get_sg_data, get_sanitized_data
+from .ui import resources_rc
+
+class ShotgunEntityModel(ShotgunModel):
+    """
+    A model that contains a hierarchy of Shotgun entity data and sets the icon for each item
+    to the icon for the entity type if available.
+
+    For Step entities, the icon will be a colour swatch based on the Step color field
+    """
+
+    # global cache of step colours - avoids querying from Shotgun multiple times!
+    _SG_STEP_COLOURS = {}
+
+    def __init__(self, entity_type, filters, hierarchy, fields, parent, 
+                 download_thumbs=False, schema_generation=0, bg_load_thumbs=False,
+                 bg_task_manager=None):
+        """
+        Construction
+
+        :param entity_type:         The type of the entities that should be loaded into this model.
+        :param filters:             A list of filters to be applied to entities in the model - these 
+                                    will be passed to the Shotgun API find() call when populating the 
+                                    model
+        :param hierarchy:           List of Shotgun fields that will be used to define the structure 
+                                    of the items in the model.
+        :param fields:              List of Shotgun fields to populate the items in the model with.
+                                    These will be passed to the Shotgun API find() call when populating
+                                    the model.
+        :param parent:              Parent QObject.
+        :param download_thumbs:     Boolean to indicate if this model should attempt
+                                    to download and process thumbnails for the downloaded data.
+        :param schema_generation:   Schema generation index. If you are changing the format
+                                    of the data you are retrieving from Shotgun, and therefore
+                                    want to invalidate any cache files that may already exist
+                                    in the system, you can increment this integer.
+        :param bg_load_thumbs:      If set to True, thumbnails will be loaded in the background.
+        :param bg_task_manager:     Background task manager to use for any asynchronous work.  If
+                                    this is None then a task manager will be created as needed.
+        """
+        self._entity_icons = {}
+        self._step_swatch_icons = {}
+
+        # make sure fields is valid:
+        fields = fields or []
+        # for backwards compatibility, make sure certain fields are added:
+        fields = list(set(fields + ["image", "sg_status_list", "description"]))
```

Comment:
not sure the comment or the code makes sense - I'll remove it and see what breaks!